cirrus-ci_env/cirrus-ci_env.py

```diff
@@ -0,0 +1,259 @@
+#!/usr/bin/env python3
+
+"""Utility to provide canonical listing of Cirrus-CI tasks and env. vars."""
+
+import argparse
+import re
+import sys
+
+import yaml
+
+
+class DefFmt(dict):
+    """
+    Defaulting-dict helper class for render_env()'s str.format_map().
+
+    See: https://docs.python.org/3.7/library/stdtypes.html#str.format_map
+    """
+
+    dollar_env_var = re.compile(r"\$(\w+)")
+    dollarcurly_env_var = re.compile(r"\$\{(\w+)\}")
+
+    def __missing__(self, key: str) -> str:
+        """Not-found items converted back to shell env var format."""
+        return "${{{0}}}".format(key)
+
+
+class CirrusCfg():
+    """Represent a fully realized list of .cirrus.yml tasks."""
+
+    # Dictionary of global, configuration-wide environment variable values.
+    global_env = None
+
+    # String values representing instance type and image name/path/uri
+    global_type = None
+    global_image = None
+
+    def __init__(self, config: dict):
+        """Create a new instance, given a parsed .cirrus.yml config object."""
+        if not isinstance(config, dict):
+            whatsit = config.__class__
+            raise TypeError(f"Expected 'config' argument to be a dictionary, not a {whatsit}")
+        # This makes a copy, doesn't touch the original
+        self.global_env = self.render_env(config.get("env", {}))
+        self.global_type, self.global_image = self.get_type_image(config)
+        self.tasks = self.render_tasks(config)
+        self.names = list(self.tasks.keys())
+        self.names.sort()
+        self.names = tuple(self.names)  # help notice attempts to modify
+
+    def render_env(self, env: dict) -> dict:
+        """
+        Repeatedly call format_env() to render out-of-order env key values.
+
+        This is a compromise vs recursion. Since substitution values may be
+        referenced while processing, and dictionary keys have no defined
+        order.  Simply provide multiple chances for the substitution to
+        occur.  On failure, a shell-compatible variable reference is simply
+        left in place.
+        """
+        # Cirrus-CI uses 10-tries according to their support
+        out = self.format_env(env, self.global_env)
+        for _ in range(9):
+            out = self.format_env(out, self.global_env)
+        return out
+
+    @staticmethod
+    def format_env(env, global_env: dict) -> dict:
```

Comment:
```suggestion
    def format_env(env, global_env: Mapping[str, Any]) -> Dict[str, Any]:
```
Would improve checking.